File: Servers/UnityInputSyncerSocketIOServer/src/input-syncer/server-instance.ts
import { InputSyncerPlayer } from './input-syncer-player';
import { InputSyncerServer } from './input-syncer-server';
import { ServerInstanceState } from './types';

export type InstanceStateChangeHandler = (
  instance: ServerInstance,
  oldState: ServerInstanceState,
  newState: ServerInstanceState,
) => void;

export class ServerInstance {
  readonly id: string;
  readonly server: InputSyncerServer;
  readonly createdAt: Date;

  private _state = ServerInstanceState.Idle;
  private _lastStateChangeTime = new Date();

  onStateChanged: InstanceStateChangeHandler = () => {};

  constructor(id: string, server: InputSyncerServer) {
    this.id = id;
    this.server = server;
    this.createdAt = new Date();

    server.onPlayerConnected = (player: InputSyncerPlayer) =>
      this.handlePlayerConnected(player);
    server.onPlayerDisconnected = (player: InputSyncerPlayer) =>
      this.handlePlayerDisconnected(player);
    server.onMatchStarted = () => this.handleMatchStarted();
    server.onMatchFinished = () => this.handleMatchFinished();
  }

  get state(): ServerInstanceState {
    return this._state;
  }

  get lastStateChangeTime(): Date {
    return this._lastStateChangeTime;
  }

  private setState(newState: ServerInstanceState): void {
    if (this._state === newState) return;

    const oldState = this._state;
    this._state = newState;
    this._lastStateChangeTime = new Date();
    this.onStateChanged(this, oldState, newState);
  }

  private handlePlayerConnected(_player: InputSyncerPlayer): void {
    if (this._state === ServerInstanceState.Idle) {
      this.setState(ServerInstanceState.WaitingForPlayers);
    }
  }

  private handlePlayerDisconnected(_player: InputSyncerPlayer): void {
    if (
      this._state === ServerInstanceState.WaitingForPlayers &&
      this.server.getPlayerCount() === 0
    ) {
      this.setState(ServerInstanceState.Idle);
    }
  }

  private handleMatchStarted(): void {
    this.setState(ServerInstanceState.InMatch);
  }

  private handleMatchFinished(): void {
    this.setState(ServerInstanceState.Finished);
  }
}

File: Packages/com.github.behrooz-arabzade.unity-input-syncer/UnityInputSyncerUTPServer/ServerInstance.cs
using System;

namespace UnityInputSyncerUTPServer
{
    public class ServerInstance
    {
        public string Id { get; }
        public ushort Port { get; }
        public ServerInstanceState State { get; private set; }
        public InputSyncerServer Server { get; }
        public DateTime CreatedAt { get; }
        public DateTime LastStateChangeTime { get; private set; }

        public event Action<ServerInstance, ServerInstanceState, ServerInstanceState> OnStateChanged;

        public ServerInstance(string id, ushort port, InputSyncerServer server)
        {
            Id = id;
            Port = port;
            Server = server;
            State = ServerInstanceState.Idle;
            CreatedAt = DateTime.UtcNow;
            LastStateChangeTime = DateTime.UtcNow;

            Server.OnPlayerConnected += HandlePlayerConnected;
            Server.OnPlayerDisconnected += HandlePlayerDisconnected;
            Server.OnMatchStarted += HandleMatchStarted;
            Server.OnMatchFinished += HandleMatchFinished;
        }

        private void SetState(ServerInstanceState newState)
        {
            if (State == newState)
                return;

            var oldState = State;
            State = newState;
            LastStateChangeTime = DateTime.UtcNow;
            OnStateChanged?.Invoke(this, oldState, newState);
        }

        private void HandlePlayerConnected(InputSyncerServerPlayer player)
        {
            if (State == ServerInstanceState.Idle)
            {
                SetState(ServerInstanceState.WaitingForPlayers);
            }
        }

        private void HandlePlayerDisconnected(InputSyncerServerPlayer player)
        {
            if (State == ServerInstanceState.WaitingForPlayers && Server.GetPlayerCount() == 0)
            {
                SetState(ServerInstanceState.Idle);
            }
        }

        private void HandleMatchStarted()
        {
            SetState(ServerInstanceState.InMatch);
        }

        private void HandleMatchFinished()
        {
            SetState(ServerInstanceState.Finished);
        }
    }
}
